src/link/MachO/Dylib.zig

```diff
@@ -71,9 +73,48 @@ pub fn closeFile(self: Dylib) void {
     }
 }
 
+fn decodeArch(cputype: macho.cpu_type_t) !std.Target.Cpu.Arch {
+    const arch: std.Target.Cpu.Arch = switch (cputype) {
+        macho.CPU_TYPE_ARM64 => .aarch64,
+        macho.CPU_TYPE_X86_64 => .x86_64,
+        else => |value| {
+            return error.UnsupportedCpuArchitecture;
+        },
+    };
+    return arch;
+}
+
 pub fn parse(self: *Dylib) !void {
     log.debug("parsing shared library '{s}'", .{self.name.?});
 
+    var library_offset: u64 = 0;
+    {
+        const fat_header = try readFatStruct(self.file.?.reader(), macho.fat_header);
+        if (fat_header.magic == macho.FAT_MAGIC) {
+            var fat_arch_index: u32 = 0;
+            while (fat_arch_index < fat_header.nfat_arch) : (fat_arch_index += 1) {
+                const fat_arch = try readFatStruct(self.file.?.reader(), macho.fat_arch);
+                // If we come across an architecture that we do not know how to handle, that's
+                // fine because we can keep looking for one that might match.
+                const lib_arch = decodeArch(fat_arch.cputype) catch |err| switch (err) {
+                    error.UnsupportedCpuArchitecture => continue,
+                    else => |e| return e,
+                };
+                if (lib_arch == self.arch.?) {
+                    // We have found a matching architecture!
+                    try self.file.?.seekTo(fat_arch.offset);
+                    library_offset = fat_arch.offset;
+                    break;
+                }
+            } else {
+                log.err("Could not find matching cpu architecture in fat library: expected {s}", .{self.arch.?});
+                return error.MismatchedCpuArchitecture;
+            }
+        } else {
+            try self.file.?.seekTo(0);
+        }
+    }
+
```

Comment:
I think you could simplify this greatly using labeled blocks like so:

```zig
const library_offset: i64 = offset: {
    const fat_header = try readFatStruct(self.file.?.reader(), macho.fat_header);
    if (fat_header.magic != macho.FAT_MAGIC) break :offset 0;

    var fat_arch_index: u32 = 0;
    while (fat_arch_index < fat_header.nfat_arch) : (fat_arch_index += 1) {
        const fat_arch = try readFatStruct(self.file.?.reader(), macho.fat_arch);
        // If we come across an architecture that we do not know how to handle, that's
        // fine because we can keep looking for one that might match.
        const lib_arch = decodeArch(fat_arch.cputype) catch |err| switch (err) {
            error.UnsupportedCpuArchitecture => continue,
            else => |e| return e,
         };
         if (lib_arch == self.arch.?) {
             // We have found a matching architecture!
             break :offset fat_arch.offset;
         }
     } else {
         log.err("Could not find matching cpu architecture in fat library: expected {s}", .{self.arch.?});
         return error.MismatchedCpuArchitecture;
     }
};
try self.file.?.seekTo(library_offset);
```

Reply:
Based on our other conversation this should be turned into a try right (and just error if we come across unsupported arch):
```zig

// If we come across an architecture that we do not know how to handle, that's
// fine because we can keep looking for one that might match.
const lib_arch = decodeArch(fat_arch.cputype) catch |err| switch (err) {
    error.UnsupportedCpuArchitecture => continue,
    else => |e| return e,
};
```